src/services/extension-registry-provider.js

```diff
@@ -19,7 +21,41 @@
             reduce = utils.reduce,
             ownKeys = utils.ownKeys,
             toArray = utils.toArray,
-            isFunction = utils.isFunction;
+            isFunction = utils.isFunction,
+            isNaN = utils.isNaN,
+            isString = utils.isString,
+            isNumber = utils.isNumber,
+            isUndefined = utils.isUndefined,
+            // TODO: externalize this validator
+            // so that a custom sorter could be provided
+            invalidVal = -9999,
+            strToUsableNum = function(val) {
+              var test = Number(val);
+              return isNaN(test) ?
+                      invalidVal :
+                      test;
+            },
+            // ensures any provided weight is usable.
+            // see the unit test for weight
+            // - numbers returned as numbers
+            // - undefined returned as -1 (explicit 0 has greater weight)
+            // - strings returned as numbers || -9999 (invalid)
+            // - all other objects returned as -9999 (invalid)
+            // this function is intentionally resilient as it is assumed
+            // extensions could be registered by various parties.  Those
+            // who use sensible weights should be able to trust sorting
+            // placement
+            toWeight = function(value) {
+              return isNaN(value) ?
+                      invalidVal :
+                      isNumber(value) ?
+                        value :
+                        isString(value) ?
+                          strToUsableNum(value) :
+                          isUndefined(value) ?
+                            -1 :
+                            invalidVal;
```

Comment:
Here is a visual from the manual test page (`demo/list-items/index.html`):

![screen shot 2017-08-16 at 3 13 33 pm](https://user-images.githubusercontent.com/280512/29380949-8f6c03e6-8295-11e7-9d3e-ceab1177568a.png)

The idea is that:

1.  Numerical/String weight is simple (`5 vs "5"`), let it sort as expected.
2. Unweighted or undefined is treated as just about zero, I gave it `-1` to give preference to explicitly set weight.
3. Negative weights should have more value than invalid weights
4. All invalid weights drop

My initial impl was much simpler, pretty close to what you suggest.  But then I threw in a bunch of unexpected values for testing & got some strange results.  At that point a more strict "follow the rules, get rewarded" type of sort algorithm seemed to be a better solution.

The assumption here is multiple extensions could be registered to an extension point, and you may not completely trust each one that is registered.

Thoughts?